File: tests/tests_format_validator/src/feature_parser.rs
use anyhow::{Context, Result};
use regex::Regex;
use std::path::{Path, PathBuf};
use std::sync::LazyLock;

/// All recognized Gherkin scenario keywords (longest prefixes first so that
/// `Scenario Outline:` and `Scenario Template:` are matched before `Scenario:`).
const SCENARIO_PREFIXES: &[&str] = &["Scenario Template:", "Scenario Outline:", "Scenario:"];

static TAG_REGEX: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"@(\w+)").unwrap());
static STEP_REGEX: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^\s*(Given|When|Then|And|But)\s+(.+)$").unwrap());

/// Check if a trimmed line starts a scenario (any keyword).
fn is_scenario_start(line: &str) -> bool {
    SCENARIO_PREFIXES.iter().any(|p| line.starts_with(p))
}

/// Strip the scenario keyword prefix from a line, returning the scenario name.
fn strip_scenario_prefix(line: &str) -> &str {
    for prefix in SCENARIO_PREFIXES {
        if let Some(rest) = line.strip_prefix(prefix) {
            return rest.trim();
        }
    }
    line
}

#[derive(Debug, Clone)]
pub struct Feature {
    pub name: String,
    pub file_path: PathBuf,
    pub tags: Vec<String>,
    pub scenarios: Vec<Scenario>,
}

#[derive(Debug, Clone)]
pub struct Scenario {
    pub name: String,
    pub tags: Vec<String>,
    pub steps: Vec<Step>,
}

#[derive(Debug, Clone)]
pub struct Step {
    pub step_type: StepType,
    pub text: String,
}

#[derive(Debug, Clone)]
pub enum StepType {
    Given,
    When,
    Then,
    And,
    But,
}

impl Scenario {
    /// Checks that the scenario has at least one When and one Then step.
    /// Returns a list of error messages for any missing mandatory step types.
    pub fn validate_mandatory_steps(&self) -> Vec<String> {
        let mut errors = Vec::new();

        let has_when = self
            .steps
            .iter()
            .any(|s| matches!(s.step_type, StepType::When));
        let has_then = self
            .steps
            .iter()
            .any(|s| matches!(s.step_type, StepType::Then));

        if !has_when {
            errors.push(format!(
                "Scenario '{}' is missing a mandatory 'When' step",
                self.name
            ));
        }
        if !has_then {
            errors.push(format!(
                "Scenario '{}' is missing a mandatory 'Then' step",
                self.name
            ));
        }

        errors
    }
}

impl Feature {
    pub fn parse_from_file(path: &Path) -> Result<Feature> {
        let content = std::fs::read_to_string(path)
            .with_context(|| format!("Failed to read feature file: {}", path.display()))?;

        let name = path
            .file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("unknown")
            .to_string();

        let mut feature = Feature {
            name,
            file_path: path.to_path_buf(),
            tags: Vec::new(),
            scenarios: Vec::new(),
        };

        let lines: Vec<&str> = content.lines().collect();
        let mut i = 0;

        // Parse feature-level tags and feature line
        while i < lines.len() {
            let line = lines[i].trim();

            if line.starts_with('@') {
                // Parse tags
                let tags = Self::parse_tags(line);
                if feature.scenarios.is_empty() {
                    // These are feature-level tags
                    feature.tags.extend(tags);
                }
            } else if line.starts_with("Feature:") {
                // Feature declaration found
                break;
            }
            i += 1;
        }

        // Parse scenarios
        while i < lines.len() {
            let line = lines[i].trim();

            if is_scenario_start(line) {
                let scenario = Self::parse_scenario(&lines, &mut i)?;
                feature.scenarios.push(scenario);
            } else {
                i += 1;
            }
        }

        Ok(feature)
    }

    fn parse_tags(line: &str) -> Vec<String> {
        TAG_REGEX
            .captures_iter(line)
            .map(|cap| cap[1].to_string())
            .collect()
    }

    fn parse_scenario(lines: &[&str], i: &mut usize) -> Result<Scenario> {
        let mut scenario_tags = Vec::new();

        // Look for tags before the scenario
        let mut tag_start = *i;
        while tag_start > 0 && lines[tag_start - 1].trim().starts_with('@') {
            tag_start -= 1;
        }

        // Collect tags
        for line in lines.iter().take(*i).skip(tag_start).map(|s| s.trim()) {
            if line.starts_with('@') {
                scenario_tags.extend(Self::parse_tags(line));
            }
        }

        // Parse scenario name
        let scenario_line = lines[*i].trim();
        let scenario_name = strip_scenario_prefix(scenario_line).to_string();

        *i += 1;

        // Parse steps
        let mut steps = Vec::new();
        while *i < lines.len() {
            let line = lines[*i].trim();

            if line.is_empty() {
                *i += 1;
                continue;
            }

            if is_scenario_start(line) || line.starts_with("Feature:") {
                // Next scenario or feature, stop parsing this scenario
                break;
            }

            // Skip table rows (Examples tables and step data tables)
            if line.starts_with("Examples:") || line.starts_with("|") {
                *i += 1;
                continue;
            }

            if let Some(step) = Self::parse_step(line) {
                steps.push(step);
            }

            *i += 1;
        }

        Ok(Scenario {
            name: scenario_name,
            tags: scenario_tags,
            steps,
        })
    }

    fn parse_step(line: &str) -> Option<Step> {
        if let Some(captures) = STEP_REGEX.captures(line) {
            let step_type = match &captures[1] {
                "Given" => StepType::Given,
                "When" => StepType::When,
                "Then" => StepType::Then,
                "And" => StepType::And,
                "But" => StepType::But,
                _ => return None,
            };

            Some(Step {
                step_type,
                text: captures[2].trim().to_string(),
            })
        } else {
            None
        }
    }

    pub fn get_all_step_texts(&self) -> Vec<String> {
        self.scenarios
            .iter()
            .flat_map(|scenario| &scenario.steps)
            .map(|step| format!("{:?} {}", step.step_type, step.text))
            .collect()
    }
}

File: tests/test_coverage_report/feature_parser.py
#!/usr/bin/env python3
"""
Feature File Parser Module

Handles parsing of Gherkin feature files to extract scenarios,
annotations, and test method mappings.
"""

import re
from pathlib import Path
from typing import Dict, List, Optional

# Pre-compile regex patterns for better performance
TEST_CASE_PATTERN = re.compile(r'TEST_CASE\s*\(\s*"([^"]+)"')
CPP_METHOD_PATTERN = re.compile(r'void\s+([a-zA-Z_][a-zA-Z0-9_]*)\s*\(')
RUST_FUNCTION_PATTERN = re.compile(r'fn\s+([a-zA-Z_][a-zA-Z0-9_]*)\s*\(')
PYTHON_TEST_PATTERN = re.compile(r'def\s+(test_[a-zA-Z_][a-zA-Z0-9_]*)\s*\(')

# All recognized Gherkin scenario keywords (longest prefixes first so that
# "Scenario Outline:" and "Scenario Template:" are matched before "Scenario:").
SCENARIO_PREFIXES = ("Scenario Template:", "Scenario Outline:", "Scenario:")


class FeatureParser:
    """Parses Gherkin feature files and extracts test information."""
    
    def __init__(self, workspace_root: Path):
        self.workspace_root = workspace_root
        self._file_cache = {}  # Cache for file contents to avoid re-reading
    
    def clear_cache(self):
        """Clear the file cache to free memory."""
        self._file_cache.clear()
    
    @staticmethod
    def _is_scenario_start(line: str) -> bool:
        """Check if a trimmed line starts a scenario (any keyword)."""
        return any(line.startswith(p) for p in SCENARIO_PREFIXES)
    
    @staticmethod
    def _strip_scenario_prefix(line: str) -> str:
        """Strip the scenario keyword prefix from a line, returning the scenario name."""
        for prefix in SCENARIO_PREFIXES:
            if line.startswith(prefix):
                return line[len(prefix):].strip()
        return line
    
    def get_feature_scenarios(self, feature_path: str) -> List[str]:
        """Extract scenario names from a feature file."""
        try:
            full_path = self.workspace_root / feature_path
            # Use cached content if available
            if str(full_path) in self._file_cache:
                content = self._file_cache[str(full_path)]
            else:
                with open(full_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                self._file_cache[str(full_path)] = content
            
            scenarios = []
            for line in content.split('\n'):
                line = line.strip()
                if self._is_scenario_start(line):
                    scenario_name = self._strip_scenario_prefix(line)
                    scenarios.append(scenario_name)
            
            return scenarios
        except Exception as e:
            print(f"Error reading feature file {feature_path}: {e}")
            return []
    
    def get_feature_scenarios_with_annotations(self, feature_path: str) -> List[Dict[str, any]]:
        """Extract scenario names and their annotations from a feature file."""
        scenarios = []
        feature_level_expected_drivers = []
        feature_level_tags = []
        
        try:
            full_path = self.workspace_root / feature_path
            # Use cached content if available
            if str(full_path) in self._file_cache:
                content = self._file_cache[str(full_path)]
            else:
                with open(full_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                self._file_cache[str(full_path)] = content
            
            lines = content.split('\n')
            current_tags = []
            in_feature_header = True
            
            for i, line in enumerate(lines):
                line = line.strip()
                
                # Collect tags (annotations starting with @)
                if line.startswith('@'):
                    tags = [tag.strip() for tag in line.split() if tag.startswith('@')]
                    current_tags.extend(tags)
                    continue
                elif line.startswith('Feature:'):
                    # Store feature-level tags
                    feature_level_tags = current_tags.copy()
                    
                    # Process feature-level expected tags (for backward compatibility)
                    for tag in current_tags:
                        if '_expected' in tag.lower():
                            # Extract driver name from expected tag (e.g., @python_expected -> python)
                            driver_name = tag.replace('@', '').replace('_expected', '').replace('_EXPECTED', '')
                            feature_level_expected_drivers.append(driver_name.lower())
                    
                    # Reset tags for scenarios
                    current_tags = []
                    in_feature_header = False
                    continue
                
                # Process scenario (Scenario, Scenario Outline, Scenario Template)
                if self._is_scenario_start(line):
                    scenario_name = self._strip_scenario_prefix(line)
                    
                    # Behavior Differences detection is now handled by the Rust validator
                    behavior_difference_info = self._extract_behavior_difference_info(current_tags)
                    
                    # Check for scenario-level expected tags
                    scenario_expected_drivers = []
                    for tag in current_tags:
                        if '_expected' in tag.lower():
                            # Extract driver name from expected tag (e.g., @odbc_expected -> odbc)
                            driver_name = tag.replace('@', '').replace('_expected', '').replace('_EXPECTED', '')
                            scenario_expected_drivers.append(driver_name.lower())
                    
                    scenario_info = {
                        'name': scenario_name,
                        'tags': current_tags.copy(),
                        'behavior_difference_info': behavior_difference_info,
                        'expected_drivers': scenario_expected_drivers,
                        'feature_level_expected_drivers': feature_level_expected_drivers,
                        'feature_level_tags': feature_level_tags
                    }
                    
                    scenarios.append(scenario_info)
                    current_tags = []  # Reset tags after processing scenario
                    continue
                
                # Skip table rows (Examples tables and step data tables).
                # Clear tags so that tags on Examples: blocks don't leak to the next scenario.
                if line.startswith('Examples:') or line.startswith('|'):
                    current_tags = []
                    continue
                
                # Reset tags if we hit a non-tag, non-scenario line
                if line and not line.startswith('#') and not line.startswith('Background:') and not in_feature_header:
                    # Reset tags if we encounter non-tag, non-scenario content
                    if not any(keyword in line for keyword in ['Given', 'When', 'Then', 'And', 'But']):
                        current_tags = []
                        
        except Exception as e:
            print(f"Error reading feature file {feature_path}: {e}")
        
        return scenarios
    
    def format_feature_name(self, feature_name: str, feature_path: str = None) -> str:
        """Extract feature name from the feature file's 'Feature:' line."""
        if feature_path:
            try:
                full_path = self.workspace_root / feature_path
                with open(full_path, 'r') as f:
                    for line in f:
                        if line.strip().startswith('Feature:'):
                            return line.replace('Feature:', '').strip()
            except Exception:
                pass
        
        # Fallback to formatted filename
        return feature_name.replace('_', ' ').title()
    
    def extract_test_methods_with_lines(self, file_path: str, scenario_names: List[str]) -> Dict[str, int]:
        """Extract test method names and their line numbers from a test file."""
        methods = {}
        
        try:
            # Use cached content if available to avoid re-reading files
            if file_path in self._file_cache:
                content = self._file_cache[file_path]
            else:
                with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
                    content = f.read()
                self._file_cache[file_path] = content
            
            lines = content.split('\n')
            
            for i, line in enumerate(lines, 1):
                line_stripped = line.strip()
                
                # Look for TEST_CASE definitions (C++)
                test_case_match = TEST_CASE_PATTERN.search(line_stripped)
                if test_case_match:
                    test_name = test_case_match.group(1)
                    # Check if this test matches any of our scenarios
                    for scenario in scenario_names:
                        if self._method_matches_scenario(test_name, scenario):
                            methods[scenario] = i
                            break
                
                # Look for C++ method definitions
                method_match = CPP_METHOD_PATTERN.search(line_stripped)
                if method_match:
                    method_name = method_match.group(1)
                    for scenario in scenario_names:
                        if self._method_matches_scenario(method_name, scenario):
                            methods[scenario] = i
                            break
                
                # Look for Rust test functions (fn test_name() after #[test])
                rust_test_match = RUST_FUNCTION_PATTERN.search(line_stripped)
                if rust_test_match:
                    method_name = rust_test_match.group(1)
                    # Check if previous line has #[test] or #[rstest] attribute (look back a few lines)
                    is_test_function = False
                    for j in range(max(0, i-5), i):  # Look back up to 5 lines
                        if j < len(lines) and ('#[test]' in lines[j] or '#[rstest]' in lines[j]):
                            is_test_function = True
                            break
                    
                    if is_test_function:
                        for scenario in scenario_names:
                            if self._method_matches_scenario(method_name, scenario):
                                methods[scenario] = i
                                break
                
                # Look for Python test functions (def test_name():)
                python_test_match = PYTHON_TEST_PATTERN.search(line_stripped)
                if python_test_match:
                    method_name = python_test_match.group(1)
                    for scenario in scenario_names:
                        if self._method_matches_scenario(method_name, scenario):
                            methods[scenario] = i
                            break
        
        except Exception as e:
            print(f"Error reading test file {file_path}: {e}")
        
        return methods
    
    def _extract_behavior_difference_info(self, tags: List[str]) -> Optional[Dict[str, str]]:
        """Behavior Differences information is now extracted from actual implementations via Rust validator.
        This method is kept for compatibility but returns None."""
        return None
    
    def _method_matches_scenario(self, method_name: str, scenario_name: str) -> bool:
        """Check if a test method name matches a scenario name using various naming conventions."""
        def normalize(s):
            return re.sub(r'[^a-zA-Z0-9]', '', s).lower()
        
        # Convert scenario to different naming conventions
        scenario_snake = re.sub(r'[^\w\s]', '', scenario_name).replace(' ', '_').lower()
        scenario_pascal = ''.join(word.capitalize() for word in re.sub(r'[^\w\s]', '', scenario_name).split())
        scenario_test_method = f"test_{scenario_snake}"
        
        method_normalized = normalize(method_name)
        
        return (method_normalized == normalize(scenario_name) or
                method_normalized == normalize(scenario_snake) or
                method_normalized == normalize(scenario_pascal) or
                method_normalized == normalize(scenario_test_method))
